clusterloader2/pkg/util/cluster.go

```diff
@@ -0,0 +1,179 @@
+limitations under the License.
+*/
+
+package util
+
+import (
+	"fmt"
+
+	"github.com/golang/glog"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	v1helper "k8s.io/kubernetes/pkg/apis/core/v1/helper"
+)
+
+const (
+	// TODO(krzysied): These constants are copies from pkg/scheduler/algorithm/well_known_labels.go.
+	// These constants should be vendored if they are extracted to separate package.
+	taintNodeNotReady    = "node.kubernetes.io/not-ready"
+	taintNodeUnreachable = "node.kubernetes.io/unreachable"
+)
+
+var (
+	// UnreachableTaintTemplate is the taint for when a node becomes unreachable.
+	unreachableTaintTemplate = &corev1.Taint{
+		Key:    taintNodeUnreachable,
+		Effect: corev1.TaintEffectNoExecute,
+	}
+
+	// NotReadyTaintTemplate is the taint for when a node is not ready for
+	// executing pods
+	notReadyTaintTemplate = &corev1.Taint{
+		Key:    taintNodeNotReady,
+		Effect: corev1.TaintEffectNoExecute,
+	}
+)
+
+// GetNodesNumber returns number of nodes in the cluster.
+func GetNodesNumber(c clientset.Interface) (int, error) {
+	nodeList, err := c.CoreV1().Nodes().List(metav1.ListOptions{})
+	if err != nil {
+		return 0, err
+	}
+	numNodes := 0
+	for i := range nodeList.Items {
+		if isNodeSchedulable(&nodeList.Items[i]) && isNodeUntainted(&nodeList.Items[i]) {
+			numNodes++
+		}
+	}
+	return numNodes, err
+}
+
+// Node is schedulable if:
+// 1) doesn't have "unschedulable" field set
+// 2) it's Ready condition is set to true
+// 3) doesn't have NetworkUnavailable condition set to true
+func isNodeSchedulable(node *corev1.Node) bool {
+	nodeReady := isNodeConditionSetAsExpected(node, corev1.NodeReady, true, false)
+	networkReady := isNodeConditionUnset(node, corev1.NodeNetworkUnavailable) ||
+		isNodeConditionSetAsExpected(node, corev1.NodeNetworkUnavailable, false, true)
+	return !node.Spec.Unschedulable && nodeReady && networkReady
+}
+
+// Test whether a fake pod can be scheduled on "node", given its current taints.
+func isNodeUntainted(node *corev1.Node) bool {
+	fakePod := &corev1.Pod{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Pod",
+			APIVersion: "v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "fake-not-scheduled",
+			Namespace: "fake-not-scheduled",
+		},
+		Spec: corev1.PodSpec{
+			Containers: []corev1.Container{
+				{
+					Name:  "fake-not-scheduled",
+					Image: "fake-not-scheduled",
+				},
+			},
+		},
+	}
+
+	fit := podToleratesNodeTaints(fakePod, node, func(t *corev1.Taint) bool {
+		// PodToleratesNodeTaints is only interested in NoSchedule and NoExecute taints.
+		return t.Effect == corev1.TaintEffectNoSchedule || t.Effect == corev1.TaintEffectNoExecute
+	})
+	return fit
+}
+
+func podToleratesNodeTaints(pod *corev1.Pod, node *corev1.Node, filter func(t *corev1.Taint) bool) bool {
+	if v1helper.TolerationsTolerateTaintsWithFilter(pod.Spec.Tolerations, node.Spec.Taints, filter) {
```

Comment:
TBH, I look into TolerationsTolerateTaintsWithFilter implementation and it looks like for every taint (if it satisfy the filer functions) and for every toleration check `toleration.ToleratesTaint(taint)`. Assuming that there is no toleration passed, this whole function does nothing...

Reply:
No - that's not true.

It iterates over all taints (which comes from node) and checks if there is a corresponding tolleration for each of them.
If there isn't (and taint will cause pods to be not scheduled there) it will return false (and that is exactly the case we want to catch - if there are any taints than).

This is exactly what we want to do.

Reply:
You are right.
I misunderstood one case.

Reply:
Done.

Reply:
So, if I understand correctly, it just loops over the taints and return false if any of them satisfy the filter function. Maybe I should just implement the loop here, so no vendor will be needed?

Reply:
I've removed the helper call.